bpf/tests/nat_test.h

```diff
@@ -0,0 +1,131 @@
+// __BPF_BUILTINS__ in conntrack_stub.h to make sure the mock library can be
+// compiled without functions in bpf/builtins.h. Undefine __BPF_BUILTINS__ and
+// include bpf/builtins.h here.
+#undef __BPF_BUILTINS__
+#include "bpf/builtins.h"
+
+#include "bpf/ctx/skb.h"
+#include "node_config.h"
+
+// To use mock library for functions in lib/conntrack.h, lib/conntrack.h must
+// be inlcude as well as mock/mock_conntrack_stub.h.
+#include "lib/conntrack.h"
+
+// Define macros like the followings to make sure the original customized
+// functions are mapped to the mock functions.
+#define ct_lookup4 mock_ct_lookup4
+#define ct_create4 mock_ct_create4
+
+// The file containing the functions to be tested must be included after
+// defining the above macros.
+#include "lib/nat.h"
+
+// Undefine the original customized functions.
+#undef ct_lookup4
+#undef ct_create4
+
+void test_snat_v4_new_mapping() {
+    int i;
+    struct __ctx_buff ctx;
+    struct ipv4_ct_tuple otuple;
+    struct ipv4_nat_entry ostate;
+    struct ipv4_nat_target target;
+
+    // Stub the helper functions to be used before making the assertion.
+    get_prandom_u32_ExpectAndReturn(0);
+    map_lookup_elem_IgnoreAndReturn(NULL);
+    map_update_elem_IgnoreAndReturn(0);
+    ktime_get_ns_ExpectAndReturn(0);
+    // rtuple is not found in the map during the first retry and is updated
+    // successfully.
+    assert(!snat_v4_new_mapping(&ctx, &otuple, &ostate, &target));
+
+    // Each time you stub a function, it will only work when the function is
+    // first called, which means you need to stub a function exactly the same
+    // times as the number of times it will be called.
+    get_prandom_u32_ExpectAndReturn(0);
+    map_lookup_elem_IgnoreAndReturn(NULL);
+    map_update_elem_IgnoreAndReturn(1);
+    map_lookup_elem_IgnoreAndReturn(NULL);
+    map_update_elem_IgnoreAndReturn(0);
+    ktime_get_ns_ExpectAndReturn(0);
+    ktime_get_ns_ExpectAndReturn(0);
+    // rtuple fails to get updated during the first retry and updated
+    // successfully during the second retry.
+    assert(!snat_v4_new_mapping(&ctx, &otuple, &ostate, &target));
+
+    for (i = 0; i < SNAT_COLLISION_RETRIES; i++) {
+        map_lookup_elem_IgnoreAndReturn(NULL);
+        map_update_elem_IgnoreAndReturn(1);
+        ktime_get_ns_ExpectAndReturn(0);
+        get_prandom_u32_ExpectAndReturn(0);
+    }
+    skb_event_output_IgnoreAndReturn(0);
+    // rtuple keeps failing to get updated until exceeding
+    // SNAT_COLLISION_RETRIES.
+    assert(snat_v4_new_mapping(&ctx, &otuple, &ostate, &target) ==
+           DROP_NAT_NO_MAPPING);
+}
+
+void test_snat_v4_track_local() {
+    struct __ctx_buff ctx;
+    struct ipv4_ct_tuple tuple;
+    struct ipv4_nat_entry state;
+    struct ipv4_nat_target target;
+
+    mock_ct_lookup4_ExpectAnyArgsAndReturn(-1);
+    // ct_lookup4 returns a negative value so snat_v4_track_local will return
```

Comment:
It would be good if you could explain under what circumstances the CT lookup returns -1 in real scenarios. Same goes to other comments.

Reply:
>Could you split this PR into two commits - one for introducing the mocking infrastructure and friends, and other for introducing NAT tests?

Done!
> It would be good if you could explain under what circumstances the CT lookup returns -1 in real scenarios. Same goes to other comments.

Done!